### ISSUE TITLE
[Feature] A generalize way to loop/iterate over an array and run Subprocess for each item in array

### ISSUE THREAD

**Issue by TylerRick** [trailblazer/trailblazer#245]
I have an Operation that creates multiple of something (the operation's input is an array of hashes) and an Operation that creates a single record of the same thing (its input is a single hash). I'd like to have a helper/macro similar to Subprocess that makes it easy to loop/iterate over an array and call the Subprocess activity for each element.

Would that be possible, and reasonable to have in Trailblazer?

## Workaround

Currently I'm working around the lack of this feature by having a step that _manually_ invokes a subprocess/activity for each element in the array. Something like this...
```ruby
  step :process_each

  # Loop through each element and delegate to Thing::Operation::Create to do most of
  # the work
  def process_each(ctx, inputs:,**)
    inputs[:things_params].map do |thing_params|
      result = Thing::Operation::Create.({
        params: thing_params,
      })
      return result unless result.success?
    end
    true
  end
```

The nice thing about this approach (I suppose) is that you have full control over things like what to pass as inputs to the subactivity, and what to do if one of the child activities fails (that's probably the biggest question mark for how to generalize this)...

The problem with this is it doesn't actually wire directly between the parent activity and the child activity, so for instance, when you do a trace, the trace only goes as deep as this `process_each` step, but it stops there and doesn't trace any of the nested activities!

## Limitation of tracing developer tool

And maybe that's all I need. If there was a way to tell the tracer to "keep tracing" for these sub-activities, then maybe I'd be happy with this "manual" approach. The problem is, we can't _assume_ that we are currently tracing, so it can't explicitly/unconditionally call `trace` instead of `SubOperation.(...)`. And it shouldn't have to (or be able to) _know_ if it's currently tracing (because that's internal/implementation details that it shouldn't know anything about), so it wouldn't make sense for it to have a conditional like `Operation.send(tracing? ? :trace_invoke : :invoke)`.

Maybe something like this??:
```ruby
  def process_each(ctx, inputs:,**)
    inputs[:things_params].map do |thing_params|
      result = continue_trace_if_tracing(Thing::Operation::Create).({
        params: thing_params,
      })
      return result unless result.success?
    end
    true
  end
```
That's not really any better either...

## Solution/API ideas

What I was hoping for (but didn't find), was something that would let us do something like:

```ruby
  steps MapToSubactivity(Thing::Operation::Create),
    input: {things_params: :array}
```

## How?

How could it do that iteration with a single step? And we wouldn't want to define a separate step for each element in the array (not that we could anyway, since at step compile time we wouldn't know what the input will be).

So would/should we actually add some kind of low-level support for an _array_ of `steps`?

I didn't see anywhere in the code that might allow iterating or creating a step that wires in multiple tasks at once.

Here's the source for `Subprocess`:
```ruby
            def Subprocess(activity, patch: {})
              activity = Patch.customize(activity, options: patch)

              {
                task:    activity,
                outputs: Hash[activity.to_h[:outputs].collect { |output| [output.semantic, output] }]
              }
            end
```

I like how simple that is. But it makes evident that the task/step API (`{task:, outputs:}`) can only accept a single task/activity. So how would we handle iteration?

Where would this live?
- [trailblazer-activity-dsl-linear](https://github.com/trailblazer/trailblazer-activity-dsl-linear)? That was my first inclination, since that's where `Subprocess` is defined, and that's still where I'm thinking it should go, but...
[trailblazer-macro](https://github.com/trailblazer/trailblazer-macro)? Since the mapping between array input and the subprocess inputs may need to be customizable (beyond what standard I/O variable mapping gives us??), maybe it would be more at home in trailblazer-macro, with the likes of `Policy::Guard` and `Model`?

### A Macro/Activity to handle iteration?

Something like this would probably work:

___

**Comment by apotonick** [comment]
Check this: https://github.com/trailblazer/trailblazer-macro/pull/35 Might be good to move the discussion here! :beer:

**Comment by apotonick** [comment]
BTW, tracing nested operations works by using the circuit-interface, not the "convenient" public interface of an OP: http://trailblazer.to/2.1/docs/activity.html#activity-internals The "tracing" happens in `flow_options`.

**Comment by apotonick** [comment]
This is implemented now in `trailblazer` 2.1.2 and the `Each()` macro: https://trailblazer.to/2.1/docs/macro#macro-each :stars:
